### PR TITLE
VIRTS-1142/Add donut hook function

### PR DESCRIPTION
**Pull request by khyberspache**
- Enable donut shellcode generation for arbitrary CSharp payloads

### CODE REVIEW

File: app/donut.py [hunk 1]
```diff
@@ -1,3 +1,27 @@
...
+
+async def donut_handler(services, args):
+    _, file_name = await services.get('file_svc').find_file_path(args.get('file'), location='payloads')
+    exe_path, donut_ext = _stage_compatible_executable(file_name)
+    shellcode = donut.create(file=exe_path)
+    _write_shellcode_to_file(shellcode, file_name)
+    os.remove(exe_path)
+    return donut_ext, donut_ext
+
+
```
Comment (privateducky) [inline app/donut.py:14]: add """ PRIVATE """ to stick with our convention here

File: app/donut.py [hunk 2]
```diff
@@ -1,3 +1,27 @@
...
+import shutil
+
+
+async def donut_handler(services, args):
+    _, file_name = await services.get('file_svc').find_file_path(args.get('file'), location='payloads')
+    exe_path, donut_ext = _stage_compatible_executable(file_name)
+    shellcode = donut.create(file=exe_path)
+    _write_shellcode_to_file(shellcode, file_name)
+    os.remove(exe_path)
+    return donut_ext, donut_ext
```
Comment (privateducky) [inline app/donut.py:12]: I don't understand this line. why return the same thing twice?

File: app/donut.py [hunk 3]
```diff
@@ -1,3 +1,27 @@
...
+import os
+import shutil
+
+
+async def donut_handler(services, args):
+    _, file_name = await services.get('file_svc').find_file_path(args.get('file'), location='payloads')
+    exe_path, donut_ext = _stage_compatible_executable(file_name)
+    shellcode = donut.create(file=exe_path)
+    _write_shellcode_to_file(shellcode, file_name)
+    os.remove(exe_path)
```
Comment (privateducky) [inline app/donut.py:11]: I don't think I like the creation and deletion of a file in the source code itself, at run-time. 

I think we can do these file writes in the temp directory with only a slight modification. or potentially in-memory through an in-memory file (StringIO in Python).

File: app/donut.py (hunk 2 above)
Comment (khyberspache) [reply app/donut.py:12]: it's returning to the `file_svc` display_name and payload_name.  Idea being we still want to allow for obfuscation in the future

Comment (khyberspache) [reply app/donut.py:12]: I'll change to just returning the payload_name

File: app/donut.py (hunk 3 above)
Comment (khyberspache) [reply app/donut.py:11]: All I'm trying to accomplish is a file rename operation (as donut needs a .exe extension to work).  I could use `os.rename()` or similar.

File: app/donut.py (hunk 2 above)
Comment (khyberspache) [reply app/donut.py:12]: Looking at it... I'd prefer to keep the flexibility:
```Python
payload, display_name = await self._operate_extension(payload, headers)
```

**Pull request merged**